{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Simulations for evaluating archaic ancestry inference"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set paths and import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run in archanc root directory\n",
    "import os\n",
    "proj_dir = os.getcwd()\n",
    "msprime_dir = proj_dir +\"/output/msprime/\"\n",
    "archie_src_dir = proj_dir + \"/src/ArchIE/\"\n",
    "archie_out_dir = proj_dir + \"/output/ArchIE/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, proj_dir + '/src/stdpopsim')\n",
    "\n",
    "from stdpopsim import homo_sapiens, models\n",
    "import msprime\n",
    "import itertools\n",
    "import random\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import math\n",
    "import cyvcf2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Classes for generating EIGENSTRAT-formatted output from the `msprime` tree sequence\n",
    "\n",
    "### Simulating MNMs\n",
    "The function `generateEigData.mnms()` will randomly select a fraction of variant sites to be MNMs. The second mutation in each MNM is randomly placed up to K bp downstream of the first, and the genotypes of the first mutation are duplicated. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class mnmEigData:\n",
    "    \"\"\"\n",
    "    object defines .geno and .snp matrices \n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, geno, snp, prefix, repeat_rows): \n",
    "        self.geno = geno\n",
    "        self.snp = snp\n",
    "        self.prefix = prefix\n",
    "        self.repeat_rows = repeat_rows\n",
    "\n",
    "\n",
    "class generateEigData:\n",
    "    \"\"\"\n",
    "    class describes the .snp, .geno, and .ind EIGENSTRAT formats\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self, ts, model_label, rep_label=0, mnm_frac=0, mnm_dist=0):\n",
    "        self.ts = ts\n",
    "        self.model_label = model_label\n",
    "        self.rep_label = rep_label\n",
    "        if mnm_frac > 0:\n",
    "            self.sim_mnms = True\n",
    "        else:\n",
    "            self.sim_mnms = False\n",
    "        self.mnm_frac = mnm_frac\n",
    "        self.mnm_dist = mnm_dist\n",
    "        self.geno = self.geno()\n",
    "        self.snp = self.snp()\n",
    "        self.prefix = self.prefix()\n",
    "    \n",
    "    def prefix(self):\n",
    "        \"\"\"\n",
    "        get unique identifier for model based on inputs\n",
    "        \"\"\"\n",
    "\n",
    "        prefix = self.model_label + \"_rep\" + str(self.rep_label)\n",
    "#         if self.sim_mnms:\n",
    "#             prefix = prefix + \"_mnm\"+str(self.mnm_dist)+\"-\"+str(self.mnm_frac)\n",
    "        \n",
    "        return prefix\n",
    "    \n",
    "    def geno(self):\n",
    "        \"\"\"\n",
    "        generate .geno file from tree sequence\n",
    "        \"\"\"\n",
    "\n",
    "        geno = np.zeros(200, dtype=np.int8)\n",
    "        \n",
    "        for variant in list(self.ts.variants()):\n",
    "            geno = np.vstack([geno, variant.genotypes])\n",
    "        \n",
    "        geno = np.delete(geno, (0), axis=0) # remove dummy 1st row\n",
    "        return geno\n",
    "                \n",
    "    def snp(self):\n",
    "        \"\"\"\n",
    "        generate .snp file from tree sequence\n",
    "        \"\"\"\n",
    "\n",
    "        d = pd.DataFrame()\n",
    "        for variant in list(self.ts.variants()):\n",
    "            d = d.append(pd.DataFrame({'ID': \"1:\"+str(round(variant.site.position)),\n",
    "                      'CHR': \"1\",\n",
    "                      'POS': str(round(variant.site.position)),\n",
    "                      'POS1': str(variant.site.position/10e6), \n",
    "                      'REF': \"A\", \n",
    "                      'ALT': \"G\"\n",
    "                     }, index=[0]), ignore_index=True)\n",
    "\n",
    "        snp_df = pd.DataFrame(d)\n",
    "        return snp_df\n",
    "    \n",
    "    def mnms(self):\n",
    "        \"\"\"\n",
    "        update .geno and .snp output to include simulated MNMs\n",
    "        \"\"\"\n",
    "        \n",
    "        snp_mnm = pd.DataFrame()\n",
    "#         geno_mat = self.geno()\n",
    "        repeat_rows = []\n",
    "        for index, snp in self.snp.iterrows():\n",
    "            snp_mnm = snp_mnm.append(snp, ignore_index=True)\n",
    "            \n",
    "            random.seed(snp['POS'])\n",
    "            if random.random() < self.mnm_frac:\n",
    "\n",
    "                dist = random.randint(1, self.mnm_dist)\n",
    "\n",
    "                mnm_snp = snp\n",
    "                mnm_snp['POS'] = str(round(int(snp['POS'])+dist))\n",
    "                mnm_snp['POS1'] = float(mnm_snp['POS'])/10e6\n",
    "                mnm_snp['ID'] = \"1:\" + mnm_snp['POS']\n",
    "                snp_mnm = snp_mnm.append(mnm_snp, ignore_index=True)\n",
    "                repeat_rows.append(2)\n",
    "            else:\n",
    "                repeat_rows.append(1)\n",
    "        \n",
    "        geno_mat_mnm = np.repeat(self.geno, repeats=repeat_rows, axis=0)\n",
    "\n",
    "        mnm_prefix = self.prefix + \"_mnm\"+str(self.mnm_dist)+\"-\"+str(self.mnm_frac)\n",
    "        \n",
    "        return mnmEigData(geno=geno_mat_mnm,\n",
    "                          snp=snp_mnm,\n",
    "                          prefix=mnm_prefix,\n",
    "                          repeat_rows=repeat_rows)\n",
    "\n",
    "    \n",
    "class writeEigData:\n",
    "    \"\"\"\n",
    "    functions for writing .snp .geno and .ind files\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self, eig_data, output_dir=\"./\"):\n",
    "        self.eig_data = eig_data\n",
    "        self.prefix = eig_data.prefix\n",
    "        self.output_dir = output_dir\n",
    "\n",
    "    def dump(self):\n",
    "        \"\"\"\n",
    "        run all 3 output functions: write_snp, write_geno, and write_ind\n",
    "        \"\"\"\n",
    "        \n",
    "        self.write_snp()\n",
    "        self.write_geno()\n",
    "        self.write_ind()\n",
    "        \n",
    "    def write_snp(self):\n",
    "        \"\"\"\n",
    "        write .snp files\n",
    "        \"\"\"\n",
    "\n",
    "        self.eig_data.snp.to_csv(self.output_dir + self.prefix + \".snp\", \n",
    "                  index=True,\n",
    "                  sep=\"\\t\")\n",
    "\n",
    "    def write_geno(self):\n",
    "        \"\"\"\n",
    "        write .geno files\n",
    "        \"\"\"\n",
    "\n",
    "        geno_afr = self.eig_data.geno[:,:100]\n",
    "        geno_eur = self.eig_data.geno[:,100:200]\n",
    "\n",
    "        geno_pop = [geno_afr, geno_eur]\n",
    "\n",
    "        for i, pop in enumerate([\"afr\", \"eur\"]):\n",
    "            np.savetxt(self.output_dir + self.prefix + \"_\" + pop + \".geno\", geno_pop[i], delimiter=\"\", fmt='%i')\n",
    "\n",
    "    def write_ind(self):\n",
    "        \"\"\"\n",
    "        write .ind files\n",
    "        \"\"\"\n",
    "\n",
    "        for pop in [\"afr\", \"eur\"]:\n",
    "            # write out separate .ind files per population.\n",
    "            # columns indicate sample ID, sex (set as 'U'), and label (set as 'ADMIXED')\n",
    "            ind_file = self.output_dir + self.prefix + \"_\" + pop + \".ind\"\n",
    "            with open(ind_file, \"w\") as id_file:\n",
    "                for sample_id in range(0,100):\n",
    "                    sample_name = self.prefix + \"_\" + pop + \\\n",
    "                        \"_sample_\" + str(sample_id)\n",
    "                    print(\"\\t\".join([sample_name, \"U\", pop]),  file=id_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simulate models\n",
    "\n",
    "Simulate 200 samples (100 each of European and African ancestry) under each of the specified models (with 1000 replicates each)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# coalescent simulation parameters\n",
    "sample_size = 100 # each\n",
    "length = 50000\n",
    "mu = 1.15e-8\n",
    "rr = 1e-8\n",
    "replicates = 1000\n",
    "seed = 30\n",
    "\n",
    "# Gutenkunst 3-population model\n",
    "GutenkunstThreePop_model = homo_sapiens.GutenkunstThreePopOutOfAfrica()\n",
    "GutenkunstThreePop_ts = msprime.simulate(\n",
    "    # first 100 samples from AFR, next 100 from EUR\n",
    "    samples=[msprime.Sample(0, 0)]*sample_size + [msprime.Sample(1, 0)]*sample_size,\n",
    "    length=length, \n",
    "    mutation_rate=mu, \n",
    "    recombination_rate=rr,\n",
    "    random_seed=seed,\n",
    "    num_replicates=replicates,\n",
    "    **GutenkunstThreePop_model.asdict())\n",
    "\n",
    "# Tennessen 2-population model\n",
    "TennessenTwoPop_model = homo_sapiens.TennessenTwoPopOutOfAfrica()\n",
    "TennessenTwoPop_ts = msprime.simulate(\n",
    "    # first 100 samples from AFR, next 100 from EUR\n",
    "    samples=[msprime.Sample(0, 0)]*sample_size + [msprime.Sample(1, 0)]*sample_size,\n",
    "    length=length, \n",
    "    mutation_rate=mu, \n",
    "    recombination_rate=rr,\n",
    "    random_seed=seed,\n",
    "    num_replicates=replicates,\n",
    "    **TennessenTwoPop_model.asdict())\n",
    "\n",
    "#-------------------------------------------------------\n",
    "# define other models here and add to model_dict below\n",
    "#-------------------------------------------------------\n",
    "\n",
    "# modify demographic parameters to include archaic branches\n",
    "# GutenkunstThreePopArchaic_model = homo_sapiens.GutenkunstThreePopArchaic()\n",
    "\n",
    "# GutenkunstThreePopArchaic_ts = msprime.simulate(\n",
    "#     # first 100 samples from AFR, next 100 from EUR\n",
    "#     samples=[msprime.Sample(0, 0)]*sample_size + [msprime.Sample(1, 0)]*sample_size,\n",
    "#     length=length, \n",
    "#     mutation_rate=mu, \n",
    "#     recombination_rate=rr,\n",
    "#     random_seed=seed,\n",
    "#     num_replicates=replicates,\n",
    "#     **GutenkunstThreePopArchaic_model.asdict())\n",
    "\n",
    "# create dictionary of models\n",
    "model_dict = {\"GutenkunstThreePop\": GutenkunstThreePop_ts,\n",
    "             \"TennessenTwoPop\": TennessenTwoPop_ts}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get simulated data and run archaic admixture detection methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "run_archie = False\n",
    "for model_label, model in model_dict.items():    \n",
    "    for j, ts in enumerate(model):\n",
    "        if j == 1: # testing with just 1st simulated tree sequence\n",
    "            \n",
    "            eig_data = generateEigData(ts, model_label, rep_label=j, mnm_frac=0.015, mnm_dist=100)\n",
    "            print(\"---Without MNMs---\")\n",
    "            print(eig_data.prefix)\n",
    "            print(eig_data.geno.shape)\n",
    "            print(eig_data.snp.shape)\n",
    "            writeEigData(eig_data, msprime_dir).dump()\n",
    "            \n",
    "            # write non-MNM data to VCF\n",
    "            ts.write_vcf(msprime_dir + eig_data.prefix + \".vcf\", 2)\n",
    "            \n",
    "            print(\"---With MNMs---\")\n",
    "            eig_data_mnms = eig_data.mnms()\n",
    "            print(eig_data_mnms.prefix)\n",
    "            print(eig_data_mnms.geno.shape)\n",
    "            print(eig_data_mnms.snp.shape)\n",
    "            writeEigData(eig_data_mnms, msprime_dir).dump()\n",
    "            \n",
    "            # add code for converting .snp/.geno/.ind data to VCF\n",
    "            # alternatively, use cyvcf2 to read the non-MNM VCF into a numpy array \n",
    "            # and duplicate rows with the repeat_rows indices\n",
    "            #\n",
    "            # old_gts = VCF genotypes as numpy array\n",
    "            # ...\n",
    "            # new_gts = np.repeat(self.geno, repeats=eig_data_mnms.repeat_rows, axis=0)\n",
    "            # ...\n",
    "            # write new VCF\n",
    "            \n",
    "            if run_archie:\n",
    "                for data in [eig_data, eig_data_mnms]:\n",
    "                    for pop in [\"afr\", \"eur\"]:\n",
    "\n",
    "                        if pop == \"afr\":\n",
    "                            ref_pop = \"eur\"\n",
    "                        else:\n",
    "                            ref_pop = \"afr\"\n",
    "\n",
    "                        prefix = data.prefix\n",
    "                        stats_pop_cmd = \"python \" + archie_src_dir + \"data/calc_stats_window_data.py\" + \\\n",
    "                            \" -s \" + msprime_dir + prefix + \".snp\" + \\\n",
    "                            \" -i \" + msprime_dir + prefix + \"_\" + pop + \".ind\" + \\\n",
    "                            \" -a \" + msprime_dir + prefix + \"_\" + pop + \".geno\" + \\\n",
    "                            \" -r \" + msprime_dir + prefix + \"_\" + ref_pop + \".geno\" + \\\n",
    "                            \" -c 1 -b 0 -e 50000 -w 50000 -z 50000 \" + \\\n",
    "                            \" > \" + archie_out_dir + prefix + \"_\" + pop + \".txt\"  \n",
    "                        print(stats_pop_cmd + \"\\n\")\n",
    "#                         os.system(stats_pop_cmd)\n",
    "            \n",
    "            # add code for evaluating additional methods\n",
    "            # if run_sprime:\n",
    "            # ...\n",
    "            # ...\n",
    "            \n",
    "            # if run_moments:\n",
    "            # ...\n",
    "            # ...\n",
    "            \n",
    "            # if run_idetect:\n",
    "            # ...\n",
    "            # ..."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "anaconda",
   "language": "python",
   "name": "anaconda"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
