{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Simulations for evaluating archaic ancestry inference"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set paths and import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run in archanc root directory\n",
    "import os\n",
    "proj_dir = os.getcwd()\n",
    "msprime_dir = proj_dir +\"/output/msprime/\"\n",
    "archie_src_dir = proj_dir + \"/src/ArchIE/\"\n",
    "archie_out_dir = proj_dir + \"/output/ArchIE/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, proj_dir + '/src/stdpopsim')\n",
    "\n",
    "from stdpopsim import homo_sapiens, models\n",
    "import msprime\n",
    "import itertools\n",
    "import random\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import math\n",
    "import cyvcf2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Classes for generating EIGENSTRAT-formatted output from the `msprime` tree sequence\n",
    "\n",
    "### Simulating MNMs\n",
    "The function `generateEigData.mnms()` will randomly select a fraction of variant sites to be MNMs. The second mutation in each MNM is randomly placed up to K bp downstream of the first, and the genotypes of the first mutation are duplicated. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "class mnmEigData:\n",
    "    \"\"\"\n",
    "    object defines .geno and .snp matrices \n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, geno, snp, prefix, repeat_rows): \n",
    "        self.geno = geno\n",
    "        self.snp = snp\n",
    "        self.prefix = prefix\n",
    "        self.repeat_rows = repeat_rows\n",
    "\n",
    "\n",
    "class generateEigData:\n",
    "    \"\"\"\n",
    "    class describes the .snp, .geno, and .ind EIGENSTRAT formats\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self, ts, model_label, rep_label=0, mnm_frac=0, mnm_dist=0):\n",
    "        self.ts = ts\n",
    "        self.model_label = model_label\n",
    "        self.rep_label = rep_label\n",
    "        if mnm_frac > 0:\n",
    "            self.sim_mnms = True\n",
    "        else:\n",
    "            self.sim_mnms = False\n",
    "        self.mnm_frac = mnm_frac\n",
    "        self.mnm_dist = mnm_dist\n",
    "        self.geno = self.geno()\n",
    "        self.snp = self.snp()\n",
    "        self.prefix = self.prefix()\n",
    "    \n",
    "    def prefix(self):\n",
    "        \"\"\"\n",
    "        get unique identifier for model based on inputs\n",
    "        \"\"\"\n",
    "\n",
    "        prefix = self.model_label + \"_rep\" + str(self.rep_label)\n",
    "#         if self.sim_mnms:\n",
    "#             prefix = prefix + \"_mnm\"+str(self.mnm_dist)+\"-\"+str(self.mnm_frac)\n",
    "        \n",
    "        return prefix\n",
    "    \n",
    "    def geno(self):\n",
    "        \"\"\"\n",
    "        generate .geno file from tree sequence\n",
    "        \"\"\"\n",
    "\n",
    "        geno = np.zeros(200, dtype=np.int8)\n",
    "        \n",
    "        for variant in list(self.ts.variants()):\n",
    "            geno = np.vstack([geno, variant.genotypes])\n",
    "        \n",
    "        geno = np.delete(geno, (0), axis=0) # remove dummy 1st row\n",
    "        return geno\n",
    "                \n",
    "    def snp(self):\n",
    "        \"\"\"\n",
    "        generate .snp file from tree sequence\n",
    "        \"\"\"\n",
    "\n",
    "        d = pd.DataFrame(columns=['ID', 'CHR', 'POS', 'POS1', 'REF', 'ALT'])\n",
    "        for variant in list(self.ts.variants()):\n",
    "            d = d.append(pd.DataFrame({'ID': \"1:\"+str(round(variant.site.position)),\n",
    "                      'CHR': \"1\",\n",
    "                      'POS': str(round(variant.site.position)),\n",
    "                      'POS1': str(variant.site.position/10e6), \n",
    "                      'REF': \"A\", \n",
    "                      'ALT': \"G\"\n",
    "                     }, index=[0]), ignore_index=True)\n",
    "\n",
    "        snp_df = pd.DataFrame(d)\n",
    "        return snp_df\n",
    "    \n",
    "    def mnms(self):\n",
    "        \"\"\"\n",
    "        update .geno and .snp output to include simulated MNMs\n",
    "        \"\"\"\n",
    "        \n",
    "        snp_mnm = pd.DataFrame(columns=['ID', 'CHR', 'POS', 'POS1', 'REF', 'ALT'])\n",
    "#         geno_mat = self.geno()\n",
    "        repeat_rows = []\n",
    "        for index, snp in self.snp.iterrows():\n",
    "            snp_mnm = snp_mnm.append(snp, ignore_index=True)\n",
    "            \n",
    "            random.seed(snp['POS'])\n",
    "            if random.random() < self.mnm_frac:\n",
    "\n",
    "                dist = random.randint(1, self.mnm_dist)\n",
    "\n",
    "                mnm_snp = snp\n",
    "                mnm_snp['POS'] = str(round(int(snp['POS'])+dist))\n",
    "                mnm_snp['POS1'] = float(mnm_snp['POS'])/10e6\n",
    "                mnm_snp['ID'] = \"1:\" + mnm_snp['POS']\n",
    "                snp_mnm = snp_mnm.append(mnm_snp, ignore_index=True)\n",
    "                repeat_rows.append(2)\n",
    "            else:\n",
    "                repeat_rows.append(1)\n",
    "        \n",
    "        geno_mat_mnm = np.repeat(self.geno, repeats=repeat_rows, axis=0)\n",
    "\n",
    "        mnm_prefix = self.prefix + \"_mnm\"+str(self.mnm_dist)+\"-\"+str(self.mnm_frac)\n",
    "        \n",
    "        return mnmEigData(geno=geno_mat_mnm,\n",
    "                          snp=snp_mnm,\n",
    "                          prefix=mnm_prefix,\n",
    "                          repeat_rows=repeat_rows)\n",
    "\n",
    "    \n",
    "class writeEigData:\n",
    "    \"\"\"\n",
    "    functions for writing .snp .geno and .ind files\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self, eig_data, output_dir=\"./\"):\n",
    "        self.eig_data = eig_data\n",
    "        self.prefix = eig_data.prefix\n",
    "        self.output_dir = output_dir\n",
    "\n",
    "    def dump(self):\n",
    "        \"\"\"\n",
    "        run all 3 output functions: write_snp, write_geno, and write_ind\n",
    "        \"\"\"\n",
    "        \n",
    "        self.write_snp()\n",
    "        self.write_geno()\n",
    "        self.write_ind()\n",
    "        \n",
    "    def write_snp(self):\n",
    "        \"\"\"\n",
    "        write .snp files\n",
    "        \"\"\"\n",
    "\n",
    "        self.eig_data.snp.to_csv(self.output_dir + self.prefix + \".snp\", \n",
    "                  index=True,\n",
    "                  sep=\"\\t\")\n",
    "\n",
    "    def write_geno(self):\n",
    "        \"\"\"\n",
    "        write .geno files\n",
    "        \"\"\"\n",
    "\n",
    "        geno_afr = self.eig_data.geno[:,:100]\n",
    "        geno_eur = self.eig_data.geno[:,100:200]\n",
    "\n",
    "        geno_pop = [geno_afr, geno_eur]\n",
    "\n",
    "        for i, pop in enumerate([\"afr\", \"eur\"]):\n",
    "            np.savetxt(self.output_dir + self.prefix + \"_\" + pop + \".geno\", geno_pop[i], delimiter=\"\", fmt='%i')\n",
    "\n",
    "    def write_ind(self):\n",
    "        \"\"\"\n",
    "        write .ind files\n",
    "        \"\"\"\n",
    "\n",
    "        for pop in [\"afr\", \"eur\"]:\n",
    "            # write out separate .ind files per population.\n",
    "            # columns indicate sample ID, sex (set as 'U'), and label (set as 'ADMIXED')\n",
    "            ind_file = self.output_dir + self.prefix + \"_\" + pop + \".ind\"\n",
    "            with open(ind_file, \"w\") as id_file:\n",
    "                for sample_id in range(0,100):\n",
    "                    sample_name = self.prefix + \"_\" + pop + \\\n",
    "                        \"_sample_\" + str(sample_id)\n",
    "                    print(\"\\t\".join([sample_name, \"U\", pop]),  file=id_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "class mnmVCFData:\n",
    "    \"\"\"\n",
    "    object defines .vcf matrices\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, vcf, prefix, repeat_rows):\n",
    "        self.vcf = vcf\n",
    "        self.prefix = prefix\n",
    "        self.repeat_rows = repeat_rows\n",
    "\n",
    "class generateVCFData:\n",
    "    \"\"\"\n",
    "    generate VCF files from tree sequence\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, ts, model_label, rep_label=0, mnm_frac=0, mnm_dist=0):\n",
    "        self.ts = ts\n",
    "        self.model_label = model_label\n",
    "        self.rep_label = rep_label\n",
    "        if mnm_frac > 0:\n",
    "            self.sim_mnms = True\n",
    "        else:\n",
    "            self.sim_mnms = False\n",
    "        self.mnm_frac = mnm_frac\n",
    "        self.mnm_dist = mnm_dist\n",
    "        self.vcf = self.vcf()\n",
    "        self.prefix = self.prefix()\n",
    "\n",
    "    def prefix(self):\n",
    "        \"\"\"\n",
    "        get unique identifier for model based on inputs\n",
    "        \"\"\"\n",
    "\n",
    "        prefix = self.model_label + \"_rep\" + str(self.rep_label)\n",
    "        return prefix\n",
    "\n",
    "\n",
    "    def vcf(self):\n",
    "        \"\"\"\n",
    "        generate data frame of genotypes in VCF format\n",
    "        \"\"\"\n",
    "\n",
    "        # for variant in self.ts.variants():\n",
    "        #     pos = round(variant.site.position)\n",
    "            # output current variant to VCF\n",
    "\n",
    "                # text_file.write(\"\\t\".join(snv) + \"\\n\")\n",
    "\n",
    "        snvs = []\n",
    "        # d = np.empty([59, 2], dtype=int)\n",
    "        for variant in list(self.ts.variants()):\n",
    "            pos = round(variant.site.position)\n",
    "            snv = [\"1\", str(pos), \"1_%s\" % pos, \"A\", \"G\", \"50\", \"PASS\", \"VT=SNP\", \"GT\"]\n",
    "            snv.extend([(\"|\").join([str(g) for g in variant.genotypes[i: i+2]]) for i in range(0, len(variant.genotypes), 2)])\n",
    "            snvs.append(snv)\n",
    "            \n",
    "#         d = d.append(pd.DataFrame(snv, index=[0]), ignore_index=True)\n",
    "        snv_df = pd.DataFrame(snvs)\n",
    "        return snv_df\n",
    "    \n",
    "    def mnms(self):\n",
    "        \"\"\"\n",
    "        update .vcf output to include simulated MNMs\n",
    "        \"\"\"\n",
    "\n",
    "        snv_df = pd.DataFrame()\n",
    "#         geno_mat = self.geno()\n",
    "        repeat_rows = []\n",
    "        for index, snv in self.vcf.iterrows():\n",
    "            snv_df.append(snv, ignore_index=True)\n",
    "\n",
    "            random.seed(int(snv[1]))\n",
    "            if random.random() < self.mnm_frac:\n",
    "                # make sure a mnm is at least 10bp from its counterpart.\n",
    "                # this is to avoid the internal filter of Sprime.\n",
    "                mnm_snv = snv\n",
    "                dist = random.randint(10, self.mnm_dist)\n",
    "                mnm_cand = int(snv[1]) + dist\n",
    "                mnm_cand_r = str(round(mnm_cand))\n",
    "                mnm_snv[1] = str(mnm_cand_r)\n",
    "                mnm_snv[2] = \"1_%s\" % mnm_cand_r\n",
    "                snv_df.append(snv_df, ignore_index=True)\n",
    "                repeat_rows.append(2)\n",
    "            else:\n",
    "                repeat_rows.append(1)\n",
    "\n",
    "        mnm_prefix = self.prefix + \"_mnm\"+str(self.mnm_dist)+\"-\"+str(self.mnm_frac)\n",
    "\n",
    "        return mnmVCFData(vcf=snv_df,\n",
    "                          prefix=mnm_prefix,\n",
    "                          repeat_rows=repeat_rows)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simulate models\n",
    "\n",
    "Simulate 200 samples (100 each of European and African ancestry) under each of the specified models (with 1000 replicates each)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# coalescent simulation parameters\n",
    "sample_size = 100 # each\n",
    "length = 50000\n",
    "mu = 1.15e-8\n",
    "rr = 1e-8\n",
    "replicates = 1000\n",
    "seed = 30\n",
    "\n",
    "# Gutenkunst 3-population model\n",
    "GutenkunstThreePop_model = homo_sapiens.GutenkunstThreePopOutOfAfrica()\n",
    "GutenkunstThreePop_ts = msprime.simulate(\n",
    "    # first 100 samples from AFR, next 100 from EUR\n",
    "    samples=[msprime.Sample(0, 0)]*sample_size + [msprime.Sample(1, 0)]*sample_size,\n",
    "    length=length, \n",
    "    mutation_rate=mu, \n",
    "    recombination_rate=rr,\n",
    "    random_seed=seed,\n",
    "    num_replicates=replicates,\n",
    "    **GutenkunstThreePop_model.asdict())\n",
    "\n",
    "# Tennessen 2-population model\n",
    "TennessenTwoPop_model = homo_sapiens.TennessenTwoPopOutOfAfrica()\n",
    "TennessenTwoPop_ts = msprime.simulate(\n",
    "    # first 100 samples from AFR, next 100 from EUR\n",
    "    samples=[msprime.Sample(0, 0)]*sample_size + [msprime.Sample(1, 0)]*sample_size,\n",
    "    length=length, \n",
    "    mutation_rate=mu, \n",
    "    recombination_rate=rr,\n",
    "    random_seed=seed,\n",
    "    num_replicates=replicates,\n",
    "    **TennessenTwoPop_model.asdict())\n",
    "\n",
    "#-------------------------------------------------------\n",
    "# define other models here and add to model_dict below\n",
    "#-------------------------------------------------------\n",
    "\n",
    "# modify demographic parameters to include archaic branches\n",
    "# GutenkunstThreePopArchaic_model = homo_sapiens.GutenkunstThreePopArchaic()\n",
    "\n",
    "# GutenkunstThreePopArchaic_ts = msprime.simulate(\n",
    "#     # first 100 samples from AFR, next 100 from EUR\n",
    "#     samples=[msprime.Sample(0, 0)]*sample_size + [msprime.Sample(1, 0)]*sample_size,\n",
    "#     length=length, \n",
    "#     mutation_rate=mu, \n",
    "#     recombination_rate=rr,\n",
    "#     random_seed=seed,\n",
    "#     num_replicates=replicates,\n",
    "#     **GutenkunstThreePopArchaic_model.asdict())\n",
    "\n",
    "# create dictionary of models\n",
    "model_dict = {\"GutenkunstThreePop\": GutenkunstThreePop_ts,\n",
    "             \"TennessenTwoPop\": TennessenTwoPop_ts}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "TennessenTwoPop_ts = msprime.simulate(\n",
    "    # first 100 samples from AFR, next 100 from EUR\n",
    "    samples=[msprime.Sample(0, 0)]*sample_size + [msprime.Sample(1, 0)]*sample_size,\n",
    "    length=length, \n",
    "    mutation_rate=mu, \n",
    "    recombination_rate=rr,\n",
    "    random_seed=seed,\n",
    "    num_replicates=replicates,\n",
    "    **TennessenTwoPop_model.asdict())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=============================\n",
      "Epoch: 0 -- 204.6 generations\n",
      "=============================\n",
      "     start     end      growth_rate |     0        1    \n",
      "   -------- --------       -------- | -------- -------- \n",
      "0 |4.32e+05 1.45e+04         0.0166 |     0        0    \n",
      "1 |5.01e+05 9.28e+03         0.0195 |     0        0    \n",
      "\n",
      "Events @ generation 204.6\n",
      "   - Migration rate change for (0, 1) to 2.5e-05\n",
      "   - Migration rate change for (1, 0) to 2.5e-05\n",
      "   - Population parameter change for 1: initial_size -> 9279.212349452768 growth_rate -> 0.00307 \n",
      "   - Population parameter change for 0: initial_size -> 432124.58438330283 growth_rate -> 0 \n",
      "=================================\n",
      "Epoch: 204.6 -- 920.0 generations\n",
      "=================================\n",
      "     start     end      growth_rate |     0        1    \n",
      "   -------- --------       -------- | -------- -------- \n",
      "0 |4.32e+05 4.32e+05              0 |     0     2.5e-05 \n",
      "1 |9.28e+03 1.03e+03        0.00307 |  2.5e-05     0    \n",
      "\n",
      "Events @ generation 920.0\n",
      "   - Migration rate change for (0, 1) to 0.00015\n",
      "   - Migration rate change for (1, 0) to 0.00015\n",
      "   - Population parameter change for 1: initial_size -> 1861 growth_rate -> 0 \n",
      "==================================\n",
      "Epoch: 920.0 -- 2040.0 generations\n",
      "==================================\n",
      "     start     end      growth_rate |     0        1    \n",
      "   -------- --------       -------- | -------- -------- \n",
      "0 |4.32e+05 4.32e+05              0 |     0     0.00015 \n",
      "1 |1.86e+03 1.86e+03              0 |  0.00015     0    \n",
      "\n",
      "Events @ generation 2040.0\n",
      "   - Mass migration: lineages move from 1 to 0 with probability 1.0\n",
      "===================================\n",
      "Epoch: 2040.0 -- 5920.0 generations\n",
      "===================================\n",
      "     start     end      growth_rate |     0        1    \n",
      "   -------- --------       -------- | -------- -------- \n",
      "0 |4.32e+05 4.32e+05              0 |     0     0.00015 \n",
      "1 |1.86e+03 1.86e+03              0 |  0.00015     0    \n",
      "\n",
      "Events @ generation 5920.0\n",
      "   - Population parameter change for 0: initial_size -> 7310 \n",
      "================================\n",
      "Epoch: 5920.0 -- inf generations\n",
      "================================\n",
      "     start     end      growth_rate |     0        1    \n",
      "   -------- --------       -------- | -------- -------- \n",
      "0 |7.31e+03 7.31e+03              0 |     0     0.00015 \n",
      "1 |1.86e+03 1.86e+03              0 |  0.00015     0    \n",
      "\n"
     ]
    }
   ],
   "source": [
    "from stdpopsim import homo_sapiens, models\n",
    "import msprime\n",
    "TennessenTwoPop_model = homo_sapiens.TennessenTwoPopOutOfAfrica()\n",
    "TennessenTwoPop_model.debug()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get simulated data and run archaic admixture detection methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---Without MNMs---\n",
      "GutenkunstThreePop_rep1\n",
      "    0      1        2   3   4   5     6       7   8    9   ...   99   100  \\\n",
      "0     1     27     1_27   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "1     1     75     1_75   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "2     1    516    1_516   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "3     1    753    1_753   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "4     1   1543   1_1543   A   G  50  PASS  VT=SNP  GT  1|0 ...   0|1  1|0   \n",
      "5     1   1697   1_1697   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "6     1   2001   1_2001   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "7     1   2427   1_2427   A   G  50  PASS  VT=SNP  GT  1|1 ...   1|1  1|1   \n",
      "8     1   2718   1_2718   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "9     1   2873   1_2873   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "10    1   2944   1_2944   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "11    1   3086   1_3086   A   G  50  PASS  VT=SNP  GT  0|0 ...   1|0  0|1   \n",
      "12    1   3204   1_3204   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "13    1   3219   1_3219   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "14    1   3240   1_3240   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "15    1   4006   1_4006   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "16    1   4158   1_4158   A   G  50  PASS  VT=SNP  GT  1|0 ...   0|0  1|0   \n",
      "17    1   4186   1_4186   A   G  50  PASS  VT=SNP  GT  0|1 ...   0|0  0|0   \n",
      "18    1   4193   1_4193   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "19    1   4288   1_4288   A   G  50  PASS  VT=SNP  GT  0|1 ...   0|0  0|0   \n",
      "20    1   5104   1_5104   A   G  50  PASS  VT=SNP  GT  1|0 ...   0|0  1|0   \n",
      "21    1   5248   1_5248   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|1  0|0   \n",
      "22    1   5374   1_5374   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "23    1   5526   1_5526   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "24    1   5560   1_5560   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "25    1   5729   1_5729   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "26    1   6107   1_6107   A   G  50  PASS  VT=SNP  GT  0|0 ...   1|0  0|0   \n",
      "27    1   6676   1_6676   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "28    1   6906   1_6906   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "29    1   6936   1_6936   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "..   ..    ...      ...  ..  ..  ..   ...     ...  ..  ... ...   ...  ...   \n",
      "164   1  42206  1_42206   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "165   1  42405  1_42405   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "166   1  42589  1_42589   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "167   1  42711  1_42711   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "168   1  43071  1_43071   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "169   1  43428  1_43428   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "170   1  43504  1_43504   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|1  0|0   \n",
      "171   1  43699  1_43699   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "172   1  44062  1_44062   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|1  0|0   \n",
      "173   1  44208  1_44208   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "174   1  44393  1_44393   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "175   1  44687  1_44687   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|1  0|0   \n",
      "176   1  44777  1_44777   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|1  0|0   \n",
      "177   1  45078  1_45078   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "178   1  45642  1_45642   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "179   1  45942  1_45942   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "180   1  46180  1_46180   A   G  50  PASS  VT=SNP  GT  0|1 ...   0|0  0|0   \n",
      "181   1  46206  1_46206   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "182   1  46841  1_46841   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "183   1  47055  1_47055   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "184   1  47187  1_47187   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "185   1  47626  1_47626   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "186   1  48030  1_48030   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "187   1  48700  1_48700   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "188   1  48805  1_48805   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "189   1  48900  1_48900   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "190   1  49138  1_49138   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|1  0|0   \n",
      "191   1  49288  1_49288   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "192   1  49372  1_49372   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|1  0|0   \n",
      "193   1  49497  1_49497   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "\n",
      "     101  102  103  104  105  106  107  108  \n",
      "0    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "1    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "2    0|0  0|0  0|1  0|0  0|0  0|0  0|0  0|0  \n",
      "3    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "4    0|1  0|1  0|1  1|1  1|0  0|0  1|0  0|0  \n",
      "5    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "6    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "7    1|1  1|1  1|1  1|1  1|1  1|1  1|1  1|1  \n",
      "8    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "9    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "10   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "11   1|0  1|0  1|0  0|0  0|1  1|1  0|1  1|1  \n",
      "12   0|0  0|0  0|0  0|0  1|0  0|0  0|0  0|0  \n",
      "13   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "14   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "15   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "16   0|0  0|0  0|1  1|0  1|0  0|0  1|0  0|0  \n",
      "17   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "18   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "19   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "20   0|0  0|0  0|1  1|0  1|0  0|0  1|0  0|0  \n",
      "21   0|1  0|1  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "22   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "23   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "24   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "25   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "26   0|0  0|0  0|0  0|0  0|0  0|0  0|1  1|0  \n",
      "27   0|0  0|0  0|1  1|0  1|0  0|0  0|0  0|0  \n",
      "28   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "29   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "..   ...  ...  ...  ...  ...  ...  ...  ...  \n",
      "164  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "165  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "166  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "167  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "168  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "169  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "170  0|1  0|1  0|0  0|1  0|0  0|0  0|0  0|0  \n",
      "171  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "172  0|1  0|1  0|0  0|1  0|0  0|0  0|0  0|0  \n",
      "173  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "174  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "175  0|1  0|1  0|0  0|1  0|0  0|0  0|0  0|0  \n",
      "176  0|1  0|1  0|0  0|1  0|0  0|0  0|0  0|0  \n",
      "177  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "178  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "179  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "180  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "181  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "182  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "183  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "184  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "185  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "186  0|0  0|0  0|0  0|1  0|0  0|0  0|0  0|0  \n",
      "187  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "188  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "189  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "190  0|1  0|1  0|0  0|1  0|0  0|0  0|0  1|0  \n",
      "191  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "192  0|1  0|1  0|0  0|1  0|0  0|0  0|0  1|0  \n",
      "193  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "\n",
      "[194 rows x 109 columns]\n",
      "---With MNMs---\n",
      "GutenkunstThreePop_rep1_mnm100-0.015\n",
      "    0      1        2   3   4   5     6       7   8    9   ...   99   100  \\\n",
      "0     1     27     1_27   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "1     1     75     1_75   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "2     1    516    1_516   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "3     1    753    1_753   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "4     1   1543   1_1543   A   G  50  PASS  VT=SNP  GT  1|0 ...   0|1  1|0   \n",
      "5     1   1697   1_1697   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "6     1   2001   1_2001   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "7     1   2427   1_2427   A   G  50  PASS  VT=SNP  GT  1|1 ...   1|1  1|1   \n",
      "8     1   2718   1_2718   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "9     1   2873   1_2873   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "10    1   2944   1_2944   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "11    1   3086   1_3086   A   G  50  PASS  VT=SNP  GT  0|0 ...   1|0  0|1   \n",
      "12    1   3204   1_3204   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "13    1   3219   1_3219   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "14    1   3240   1_3240   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "15    1   4006   1_4006   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "16    1   4158   1_4158   A   G  50  PASS  VT=SNP  GT  1|0 ...   0|0  1|0   \n",
      "17    1   4186   1_4186   A   G  50  PASS  VT=SNP  GT  0|1 ...   0|0  0|0   \n",
      "18    1   4193   1_4193   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "19    1   4288   1_4288   A   G  50  PASS  VT=SNP  GT  0|1 ...   0|0  0|0   \n",
      "20    1   5104   1_5104   A   G  50  PASS  VT=SNP  GT  1|0 ...   0|0  1|0   \n",
      "21    1   5248   1_5248   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|1  0|0   \n",
      "22    1   5374   1_5374   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "23    1   5526   1_5526   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "24    1   5560   1_5560   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "25    1   5729   1_5729   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "26    1   6107   1_6107   A   G  50  PASS  VT=SNP  GT  0|0 ...   1|0  0|0   \n",
      "27    1   6676   1_6676   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "28    1   6906   1_6906   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "29    1   6936   1_6936   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "..   ..    ...      ...  ..  ..  ..   ...     ...  ..  ... ...   ...  ...   \n",
      "164   1  42206  1_42206   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "165   1  42405  1_42405   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "166   1  42589  1_42589   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "167   1  42711  1_42711   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "168   1  43071  1_43071   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "169   1  43428  1_43428   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "170   1  43504  1_43504   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|1  0|0   \n",
      "171   1  43699  1_43699   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "172   1  44062  1_44062   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|1  0|0   \n",
      "173   1  44208  1_44208   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "174   1  44393  1_44393   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "175   1  44687  1_44687   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|1  0|0   \n",
      "176   1  44777  1_44777   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|1  0|0   \n",
      "177   1  45078  1_45078   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "178   1  45642  1_45642   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "179   1  45942  1_45942   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "180   1  46180  1_46180   A   G  50  PASS  VT=SNP  GT  0|1 ...   0|0  0|0   \n",
      "181   1  46206  1_46206   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "182   1  46841  1_46841   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "183   1  47055  1_47055   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "184   1  47187  1_47187   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "185   1  47626  1_47626   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "186   1  48030  1_48030   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "187   1  48700  1_48700   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "188   1  48805  1_48805   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "189   1  48900  1_48900   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "190   1  49138  1_49138   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|1  0|0   \n",
      "191   1  49288  1_49288   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "192   1  49372  1_49372   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|1  0|0   \n",
      "193   1  49497  1_49497   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "\n",
      "     101  102  103  104  105  106  107  108  \n",
      "0    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "1    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "2    0|0  0|0  0|1  0|0  0|0  0|0  0|0  0|0  \n",
      "3    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "4    0|1  0|1  0|1  1|1  1|0  0|0  1|0  0|0  \n",
      "5    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "6    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "7    1|1  1|1  1|1  1|1  1|1  1|1  1|1  1|1  \n",
      "8    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "9    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "10   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "11   1|0  1|0  1|0  0|0  0|1  1|1  0|1  1|1  \n",
      "12   0|0  0|0  0|0  0|0  1|0  0|0  0|0  0|0  \n",
      "13   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "14   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "15   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "16   0|0  0|0  0|1  1|0  1|0  0|0  1|0  0|0  \n",
      "17   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "18   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "19   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "20   0|0  0|0  0|1  1|0  1|0  0|0  1|0  0|0  \n",
      "21   0|1  0|1  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "22   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "23   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "24   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "25   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "26   0|0  0|0  0|0  0|0  0|0  0|0  0|1  1|0  \n",
      "27   0|0  0|0  0|1  1|0  1|0  0|0  0|0  0|0  \n",
      "28   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "29   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "..   ...  ...  ...  ...  ...  ...  ...  ...  \n",
      "164  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "165  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "166  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "167  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "168  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "169  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "170  0|1  0|1  0|0  0|1  0|0  0|0  0|0  0|0  \n",
      "171  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "172  0|1  0|1  0|0  0|1  0|0  0|0  0|0  0|0  \n",
      "173  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "174  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "175  0|1  0|1  0|0  0|1  0|0  0|0  0|0  0|0  \n",
      "176  0|1  0|1  0|0  0|1  0|0  0|0  0|0  0|0  \n",
      "177  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "178  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "179  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "180  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "181  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "182  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "183  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "184  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "185  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "186  0|0  0|0  0|0  0|1  0|0  0|0  0|0  0|0  \n",
      "187  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "188  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "189  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "190  0|1  0|1  0|0  0|1  0|0  0|0  0|0  1|0  \n",
      "191  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "192  0|1  0|1  0|0  0|1  0|0  0|0  0|0  1|0  \n",
      "193  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "\n",
      "[194 rows x 109 columns]\n",
      "---Without MNMs---\n",
      "TennessenTwoPop_rep1\n",
      "    0      1        2   3   4   5     6       7   8    9   ...   99   100  \\\n",
      "0     1     90     1_90   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "1     1    122    1_122   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "2     1    150    1_150   A   G  50  PASS  VT=SNP  GT  1|0 ...   0|0  0|0   \n",
      "3     1    150    1_150   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "4     1    223    1_223   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "5     1    381    1_381   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "6     1    666    1_666   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "7     1    807    1_807   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "8     1    898    1_898   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "9     1   1244   1_1244   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "10    1   1422   1_1422   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "11    1   1438   1_1438   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "12    1   1961   1_1961   A   G  50  PASS  VT=SNP  GT  1|0 ...   0|0  0|0   \n",
      "13    1   2123   1_2123   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "14    1   2355   1_2355   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "15    1   2427   1_2427   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "16    1   2641   1_2641   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "17    1   2710   1_2710   A   G  50  PASS  VT=SNP  GT  1|0 ...   0|0  0|0   \n",
      "18    1   2798   1_2798   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  1|0   \n",
      "19    1   2935   1_2935   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "20    1   2996   1_2996   A   G  50  PASS  VT=SNP  GT  0|1 ...   0|0  1|1   \n",
      "21    1   3106   1_3106   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "22    1   3128   1_3128   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "23    1   3280   1_3280   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "24    1   3611   1_3611   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "25    1   3852   1_3852   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "26    1   4275   1_4275   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "27    1   4289   1_4289   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "28    1   4369   1_4369   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "29    1   4433   1_4433   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "..   ..    ...      ...  ..  ..  ..   ...     ...  ..  ... ...   ...  ...   \n",
      "340   1  45907  1_45907   A   G  50  PASS  VT=SNP  GT  1|0 ...   0|0  0|0   \n",
      "341   1  45984  1_45984   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "342   1  46189  1_46189   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "343   1  46317  1_46317   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "344   1  46336  1_46336   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "345   1  46593  1_46593   A   G  50  PASS  VT=SNP  GT  1|0 ...   0|0  0|0   \n",
      "346   1  46718  1_46718   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "347   1  46826  1_46826   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "348   1  46878  1_46878   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "349   1  46988  1_46988   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "350   1  47025  1_47025   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "351   1  47038  1_47038   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "352   1  47144  1_47144   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "353   1  47294  1_47294   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "354   1  47328  1_47328   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "355   1  47408  1_47408   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "356   1  47468  1_47468   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "357   1  47810  1_47810   A   G  50  PASS  VT=SNP  GT  0|0 ...   1|0  0|0   \n",
      "358   1  47915  1_47915   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "359   1  48023  1_48023   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "360   1  48098  1_48098   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "361   1  48137  1_48137   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "362   1  48179  1_48179   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "363   1  48196  1_48196   A   G  50  PASS  VT=SNP  GT  0|1 ...   0|0  1|0   \n",
      "364   1  48274  1_48274   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "365   1  48524  1_48524   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "366   1  48891  1_48891   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "367   1  49126  1_49126   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "368   1  49327  1_49327   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "369   1  49627  1_49627   A   G  50  PASS  VT=SNP  GT  1|1 ...   1|0  1|1   \n",
      "\n",
      "     101  102  103  104  105  106  107  108  \n",
      "0    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "1    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "2    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "3    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "4    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "5    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "6    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "7    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "8    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "9    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "10   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "11   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "12   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "13   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "14   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "15   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "16   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "17   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "18   1|0  0|0  0|0  0|0  1|0  0|0  0|0  0|0  \n",
      "19   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "20   1|1  1|1  1|1  0|0  1|0  0|0  1|0  1|1  \n",
      "21   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "22   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "23   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "24   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "25   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "26   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "27   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "28   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "29   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "..   ...  ...  ...  ...  ...  ...  ...  ...  \n",
      "340  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "341  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "342  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "343  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "344  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "345  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "346  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "347  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "348  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "349  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "350  0|0  0|0  0|0  0|0  0|0  0|0  0|0  1|0  \n",
      "351  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "352  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "353  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "354  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "355  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "356  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "357  0|0  0|0  0|0  0|0  0|0  0|0  0|1  0|0  \n",
      "358  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "359  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "360  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "361  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "362  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "363  1|0  0|0  0|1  1|1  0|1  1|1  0|0  0|0  \n",
      "364  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "365  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "366  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "367  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "368  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "369  1|1  1|1  1|1  1|1  1|1  1|1  1|1  0|1  \n",
      "\n",
      "[370 rows x 109 columns]\n",
      "---With MNMs---\n",
      "TennessenTwoPop_rep1_mnm100-0.015\n",
      "    0      1        2   3   4   5     6       7   8    9   ...   99   100  \\\n",
      "0     1     90     1_90   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "1     1    122    1_122   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "2     1    150    1_150   A   G  50  PASS  VT=SNP  GT  1|0 ...   0|0  0|0   \n",
      "3     1    150    1_150   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "4     1    223    1_223   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "5     1    381    1_381   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "6     1    666    1_666   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "7     1    807    1_807   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "8     1    898    1_898   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "9     1   1244   1_1244   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "10    1   1422   1_1422   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "11    1   1438   1_1438   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "12    1   1961   1_1961   A   G  50  PASS  VT=SNP  GT  1|0 ...   0|0  0|0   \n",
      "13    1   2123   1_2123   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "14    1   2355   1_2355   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "15    1   2427   1_2427   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "16    1   2641   1_2641   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "17    1   2710   1_2710   A   G  50  PASS  VT=SNP  GT  1|0 ...   0|0  0|0   \n",
      "18    1   2798   1_2798   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  1|0   \n",
      "19    1   2935   1_2935   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "20    1   2996   1_2996   A   G  50  PASS  VT=SNP  GT  0|1 ...   0|0  1|1   \n",
      "21    1   3106   1_3106   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "22    1   3128   1_3128   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "23    1   3280   1_3280   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "24    1   3611   1_3611   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "25    1   3852   1_3852   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "26    1   4275   1_4275   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "27    1   4289   1_4289   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "28    1   4369   1_4369   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "29    1   4433   1_4433   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "..   ..    ...      ...  ..  ..  ..   ...     ...  ..  ... ...   ...  ...   \n",
      "340   1  45907  1_45907   A   G  50  PASS  VT=SNP  GT  1|0 ...   0|0  0|0   \n",
      "341   1  45984  1_45984   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "342   1  46189  1_46189   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "343   1  46317  1_46317   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "344   1  46336  1_46336   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "345   1  46593  1_46593   A   G  50  PASS  VT=SNP  GT  1|0 ...   0|0  0|0   \n",
      "346   1  46718  1_46718   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "347   1  46826  1_46826   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "348   1  46878  1_46878   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "349   1  46988  1_46988   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "350   1  47025  1_47025   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "351   1  47038  1_47038   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "352   1  47144  1_47144   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "353   1  47294  1_47294   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "354   1  47328  1_47328   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "355   1  47408  1_47408   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "356   1  47514  1_47514   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "357   1  47810  1_47810   A   G  50  PASS  VT=SNP  GT  0|0 ...   1|0  0|0   \n",
      "358   1  47915  1_47915   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "359   1  48023  1_48023   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "360   1  48098  1_48098   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "361   1  48137  1_48137   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "362   1  48179  1_48179   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "363   1  48196  1_48196   A   G  50  PASS  VT=SNP  GT  0|1 ...   0|0  1|0   \n",
      "364   1  48274  1_48274   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "365   1  48524  1_48524   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "366   1  48891  1_48891   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "367   1  49126  1_49126   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "368   1  49327  1_49327   A   G  50  PASS  VT=SNP  GT  0|0 ...   0|0  0|0   \n",
      "369   1  49627  1_49627   A   G  50  PASS  VT=SNP  GT  1|1 ...   1|0  1|1   \n",
      "\n",
      "     101  102  103  104  105  106  107  108  \n",
      "0    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "1    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "2    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "3    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "4    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "5    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "6    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "7    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "8    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "9    0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "10   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "11   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "12   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "13   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "14   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "15   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "16   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "17   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "18   1|0  0|0  0|0  0|0  1|0  0|0  0|0  0|0  \n",
      "19   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "20   1|1  1|1  1|1  0|0  1|0  0|0  1|0  1|1  \n",
      "21   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "22   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "23   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "24   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "25   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "26   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "27   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "28   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "29   0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "..   ...  ...  ...  ...  ...  ...  ...  ...  \n",
      "340  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "341  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "342  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "343  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "344  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "345  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "346  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "347  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "348  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "349  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "350  0|0  0|0  0|0  0|0  0|0  0|0  0|0  1|0  \n",
      "351  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "352  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "353  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "354  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "355  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "356  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "357  0|0  0|0  0|0  0|0  0|0  0|0  0|1  0|0  \n",
      "358  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "359  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "360  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "361  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "362  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "363  1|0  0|0  0|1  1|1  0|1  1|1  0|0  0|0  \n",
      "364  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "365  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "366  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "367  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "368  0|0  0|0  0|0  0|0  0|0  0|0  0|0  0|0  \n",
      "369  1|1  1|1  1|1  1|1  1|1  1|1  1|1  0|1  \n",
      "\n",
      "[370 rows x 109 columns]\n"
     ]
    }
   ],
   "source": [
    "run_archie = False\n",
    "for model_label, model in model_dict.items():    \n",
    "    for j, ts in enumerate(model):\n",
    "        if j == 1: # testing with just 1st simulated tree sequence\n",
    "            \n",
    "#             eig_data = generateEigData(ts, model_label, rep_label=j, mnm_frac=0.015, mnm_dist=100)\n",
    "            eig_data = generateVCFData(ts, model_label, rep_label=j, mnm_frac=0.015, mnm_dist=100)\n",
    "            print(\"---Without MNMs---\")\n",
    "            print(eig_data.prefix)\n",
    "#             print(eig_data.geno.shape)\n",
    "#             print(eig_data.snp.shape)\n",
    "            print(eig_data.vcf)\n",
    "#             writeEigData(eig_data, msprime_dir).dump()\n",
    "            \n",
    "            # write non-MNM data to VCF\n",
    "#             ts.write_vcf(msprime_dir + eig_data.prefix + \".vcf\", 2)\n",
    "            \n",
    "            print(\"---With MNMs---\")\n",
    "            eig_data_mnms = eig_data.mnms()\n",
    "            print(eig_data_mnms.prefix)\n",
    "#             print(eig_data_mnms.geno.shape)\n",
    "#             print(eig_data_mnms.snp.shape)\n",
    "            print(eig_data.vcf)\n",
    "#             writeEigData(eig_data_mnms, msprime_dir).dump()\n",
    "            \n",
    "            # add code for converting .snp/.geno/.ind data to VCF\n",
    "            # alternatively, use cyvcf2 to read the non-MNM VCF into a numpy array \n",
    "            # and duplicate rows with the repeat_rows indices\n",
    "            #\n",
    "            # old_gts = VCF genotypes as numpy array\n",
    "            # ...\n",
    "            # new_gts = np.repeat(self.geno, repeats=eig_data_mnms.repeat_rows, axis=0)\n",
    "            # ...\n",
    "            # write new VCF\n",
    "            \n",
    "            if run_archie:\n",
    "                for data in [eig_data, eig_data_mnms]:\n",
    "                    for pop in [\"afr\", \"eur\"]:\n",
    "\n",
    "                        if pop == \"afr\":\n",
    "                            ref_pop = \"eur\"\n",
    "                        else:\n",
    "                            ref_pop = \"afr\"\n",
    "\n",
    "                        prefix = data.prefix\n",
    "                        stats_pop_cmd = \"python \" + archie_src_dir + \"data/calc_stats_window_data.py\" + \\\n",
    "                            \" -s \" + msprime_dir + prefix + \".snp\" + \\\n",
    "                            \" -i \" + msprime_dir + prefix + \"_\" + pop + \".ind\" + \\\n",
    "                            \" -a \" + msprime_dir + prefix + \"_\" + pop + \".geno\" + \\\n",
    "                            \" -r \" + msprime_dir + prefix + \"_\" + ref_pop + \".geno\" + \\\n",
    "                            \" -c 1 -b 0 -e 50000 -w 50000 -z 50000 \" + \\\n",
    "                            \" > \" + archie_out_dir + prefix + \"_\" + pop + \".txt\"  \n",
    "                        print(stats_pop_cmd + \"\\n\")\n",
    "#                         os.system(stats_pop_cmd)\n",
    "            \n",
    "            # add code for evaluating additional methods\n",
    "            # if run_sprime:\n",
    "            # ...\n",
    "            # ...\n",
    "            \n",
    "            # if run_moments:\n",
    "            # ...\n",
    "            # ...\n",
    "            \n",
    "            # if run_idetect:\n",
    "            # ...\n",
    "            # ..."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "anaconda",
   "language": "python",
   "name": "anaconda"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
